core/src/main/scala/org/apache/spark/internal/config/ConfigEntry.scala

```diff
@@ -89,14 +89,16 @@ private[spark] abstract class ConfigEntry[T] (
   def defaultValueString: String
 
   protected def readString(reader: ConfigReader): Option[String] = {
-    val values = Seq(
-      prependedKey.flatMap(reader.get(_)),
-      alternatives.foldLeft(reader.get(key))((res, nextKey) => res.orElse(reader.get(nextKey)))
-    ).flatten
-    if (values.nonEmpty) {
-      Some(values.mkString(prependSeparator))
-    } else {
-      None
+    // SPARK-48678: performance optimization: this code could be expressed more succinctly
+    // using flatten and mkString, but doing so adds lots of Scala collections perf. overhead.
+    val maybePrependedValue: Option[String] = prependedKey.flatMap(reader.get)
+    val value: Option[String] = alternatives
+      .foldLeft(reader.get(key))((res, nextKey) => res.orElse(reader.get(nextKey)))
+    (maybePrependedValue, value) match {
+      case (Some(prependedValue), Some(value)) => Some(s"$prependedValue$prependSeparator$value")
+      case (Some(prepend), None) => Some(prepend)
+      case (None, Some(value)) => Some(value)
```

Comment:
super nit: avoid shadowing `value` ? ... that would also allow us to simply return `value` for case 3 (and `maybePrependedValue` for case 2)

Reply:
Good idea.

I found that we can actually go a bit further and collapse all but the first case into a single case: if we don't match `(Some(prependedValue), Some(value))` then we can just return `value.orElse(prependedValue)` since only one of the two options will be non-empty in that branch. 

https://github.com/apache/spark/pull/47049/commits/a18177ff323d11db8cac2bb994e767951adcc3b1